source/MaterialXCore/MaterialNode.cpp

```diff
@@ -0,0 +1,151 @@
+    if (!parent)
+    {
+        throw Exception("Could not find a parent for material node '" + (materialNode ? materialNode->getNamePath() : EMPTY_STRING) + "'");
+    }
+
+    std::list<NodePtr> shaderNodes;
+
+    std::vector<InputPtr> inputs = materialNode->getActiveInputs();
+    if (inputs.empty())
+    {
+        // Try to find material nodes in the implementation graph if any.
+        // If a target is specified the nodedef for the given target is searched for.
+        NodeDefPtr materialNodeDef = materialNode->getNodeDef(target);
+        if (materialNodeDef)
+        {
+            InterfaceElementPtr impl = materialNodeDef->getImplementation();
+            if (impl->isA<NodeGraph>())
+            {
+                NodeGraphPtr implGraph = impl->asA<NodeGraph>();
+                for (auto defOutput : materialNodeDef->getOutputs())
+                {
+                    if (defOutput->getType() == MATERIAL_TYPE_STRING)
+                    {
+                        OutputPtr implGraphOutput = implGraph->getOutput(defOutput->getName());
+                        for (GraphIterator it = implGraphOutput->traverseGraph().begin(); it != GraphIterator::end(); ++it)
+                        {
+                            ElementPtr upstreamElem = it.getUpstreamElement();
+                            if (!upstreamElem)
+                            {
+                                it.setPruneSubgraph(true);
+                                continue;
+                            }
+                            NodePtr upstreamNode = upstreamElem->asA<Node>();
+                            if (upstreamNode && upstreamNode->getType() == MATERIAL_TYPE_STRING)
+                            {
+                                std::list<NodePtr> newShaderNodes = getShaderNodes(upstreamNode, shaderType, target);
+                                if (!newShaderNodes.empty())
+                                {
+                                    shaderNodes.insert(shaderNodes.end(), newShaderNodes.begin(), newShaderNodes.end());
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    for (const InputPtr& input : inputs) {
+        const string& inputShader = input->getNodeName();
+        if (!inputShader.empty())
+        {
+            NodePtr shaderNode = parent->getChildOfType<Node>(inputShader);
+            if (shaderNode)
+            {
+                if (!target.empty())
+                {
+                    // Look the shader definition, optionally filtered by target
+                    NodeDefPtr nodeDef = shaderNode->getNodeDef(target);
+                    if (!nodeDef)
+                    {
+                        continue;
+                    }
+                }
+                if (shaderType.empty() || shaderNode->getType() == shaderType)
```

Comment:
If no shader type is requested, here it's assumed that all inputs on the material node is of a shader type. Currently that is true, but I'm not sure it will always be the case. To be 100% safe you could check that the input really is of a type that is a shader. Could do that first thing inside the for loop.

Reply:
I will probably leave this and instead change the semantic meaning to indicate this method will return the upstream nodes connected to the inputs. If it's not specifically a shader node it can still work.